<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of gb2faa</title>
  <meta name="keywords" content="gb2faa">
  <meta name="description" content="Parse a protein FASTA file from GenBank file">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../../index.html">Home</a> &gt;  <a href="#">src</a> &gt; <a href="index.html">Utilities</a> &gt; gb2faa.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../../index.html"><img alt="<" border="0" src="../../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for src/Utilities&nbsp;<img alt=">" border="0" src="../../right.png"></a></td></tr></table>-->

<h1>gb2faa
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>Parse a protein FASTA file from GenBank file</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>function faa = gb2faa(gbFileName,ifDropNoSeqGene,faaFileName) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre class="comment"> Parse a protein FASTA file from GenBank file
 
 USAGE:
 
   faa = gb2faa(gbFileName)
   faa = gb2faa(gbFileName,ifDropNoSeqGene)
   faa = gb2faa(gbFileName,ifDropNoSeqGene,faaFileName)
   
 
 INPUTS:
   gbFileName: Genbank file name. May also include its path
 
 OPTIONAL INPUTS: 
   ifDropNoSeqGene: if genes without sequence recorded are passed. This is
                    recommended because faa with empty seq may trigger
                    errors downstream. Default: true
   faaFileName:     Output faa file name. Faa file will not be created if
                    this is left blank
 
 OUTPUTS: 
   faa: faa output format in N*1 cell. Odd entries are gene headings
        (attributes) starting with '&gt;', even entries are gene sequences. 
 
 EXAMPLE:
 
   faa = gb2faa('GCF_000005845.2_ASM584v2_genomic.gbff.tar',true,'GCF_000005845.2_ASM584v2');
 
 .. AUTHOR: - Herbert Yao, Dec 2023</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../../matlabicon.gif)">
</ul>
This function is called by:
<ul style="list-style-image:url(../../matlabicon.gif)">
</ul>
<!-- crossreference -->




<hr><address>Generated on Sun 24-Dec-2023 21:51:47 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>